File: android-app/app/src/main/java/com/musketeer/scratchpaper/activity/EditPaperActivity.java
package com.musketeer.scratchpaper.activity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.v4.widget.DrawerLayout;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import com.musketeer.baselibrary.Activity.BaseActivity;
import com.musketeer.baselibrary.util.TimeUtils;
import com.musketeer.scratchpaper.MainApplication;
import com.musketeer.scratchpaper.R;
import com.musketeer.scratchpaper.adapter.PaperListAdapter;
import com.musketeer.scratchpaper.paperfile.PaperFileUtils;
import com.musketeer.scratchpaper.utils.AppPreferenceUtils;
import com.musketeer.scratchpaper.utils.FileUtils;
import com.musketeer.scratchpaper.view.ScratchPaperView;
import com.musketeer.scratchpaper.view.ScratchPaperView.DrawStroke;
import com.umeng.analytics.MobclickAgent;

import java.util.Calendar;
import java.util.List;

public class EditPaperActivity extends BaseActivity implements OnItemClickListener{

	private static final String KEY_STORE_BITMAP="store_bitmap";
	private static final String KEY_STORE_STROKE="store_stroke";
	
	private MenuItem mPaint;
	
	private ScratchPaperView mScratchPaper;
	private DrawerLayout mDrawerLayout;
	
	private AlertDialog mDialog;
	private AlertDialog mLoadingDialog;
	private TextView loadingText;
	
    //save attribute
    private String paper_name;
    
    private ListView mSavedPaperList;
	private PaperListAdapter mAdapter;
	private List<String> mPaperList;
    
    private Handler handler =new Handler(){
    	@Override
    	//当有消息发送出来的时候就执行Handler的这个方法
    	public void handleMessage(Message msg){
    		super.handleMessage(msg);
    		//只要执行到这里就关闭对话框
    		refreshViews();
    		dismissLoadingDialog();
    		showCustomToast(getResources().getString(R.string.save_success));
    	}
	};
	
	private Handler ChangePaperHandler =new Handler(){
    	@Override
    	//当有消息发送出来的时候就执行Handler的这个方法
    	public void handleMessage(Message msg){
    		super.handleMessage(msg);
    		//只要执行到这里就关闭对话框
    		refreshViews();
    		if (paper_name!=null&&paper_name.length()>0) {
    			initPaperContent(paper_name);
    		} else {
    			mScratchPaper.setPaperAndDesk(AppPreferenceUtils.getPaperChoose(EditPaperActivity.this),
    					AppPreferenceUtils.getDeskChoose(EditPaperActivity.this));
    			mScratchPaper.setMax_undo(AppPreferenceUtils.getMaxUndo(EditPaperActivity.this));
    			mScratchPaper.clearStrokeList();
    			paper_name=TimeUtils.getDateByFileName(Calendar.getInstance().getTimeInMillis());
    		}
    		dismissLoadingDialog();
    		mDrawerLayout.closeDrawer(Gravity.LEFT);
    		showCustomToast(getResources().getString(R.string.save_success));
    	}
	};

	@Override
	public void setContentView(Bundle savedInstanceState) {
		AlertDialog.Builder builder = new AlertDialog.Builder(this);
		View view = LayoutInflater.from(this).inflate(R.layout.dialog_loading, null);
		loadingText = (TextView) view.findViewById(R.id.loading_text);
		builder.setView(view);
		mLoadingDialog = builder.create();
		setContentView(R.layout.activity_edit_paper);
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		// Inflate the menu; this adds items to the action bar if it is present.
		getSupportActionBar().setDisplayHomeAsUpEnabled(true);
		getSupportActionBar().setIcon(R.mipmap.icon_small);
		getMenuInflater().inflate(R.menu.edit_scratch_paper, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		// Handle action bar item clicks here. The action bar will
		// automatically handle clicks on the Home/Up button, so long
		// as you specify a parent activity in AndroidManifest.xml.
		AlertDialog.Builder builder;
		switch (item.getItemId()) {
			case R.id.clear_all_title:
				if (mDialog!=null) {
					mDialog.dismiss();
				}
				builder = new AlertDialog.Builder(this);
				builder.setTitle(getResources().getString(R.string.clear_all_title))
						.setMessage(getResources().getString(R.string.affirm_clear_all));
				builder.setNegativeButton(getResources().getString(R.string.no), new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						mDialog.dismiss();
					}
				});
				builder.setPositiveButton(getResources().getString(R.string.yes), new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						mScratchPaper.clearAll();
						mDialog.dismiss();
					}
				});
				mDialog = builder.create();
				mDialog.show();
				break;
			case R.id.full_screen:
				toggleFullScreen();
				break;
			case R.id.add_new_paper:
				changePaperContent(null);
				break;
			case R.id.undo:
				mScratchPaper.undoLastAction();
				break;
			case R.id.paint:
				mPaint=item;
				break;
			case R.id.black:
				mPaint.setIcon(R.mipmap.paint_setting_black);
				mScratchPaper.setColor(Color.BLACK);
				mScratchPaper.setStrokeWidth(5);
				mScratchPaper.setIsPointNotice(false);
				break;
			case R.id.red:
				mPaint.setIcon(R.mipmap.paint_setting_red);
				mScratchPaper.setColor(Color.RED);
				mScratchPaper.setStrokeWidth(5);
				mScratchPaper.setIsPointNotice(false);
				break;
			case R.id.blue:
				mPaint.setIcon(R.mipmap.paint_setting_blue);
				mScratchPaper.setColor(Color.BLUE);
				mScratchPaper.setStrokeWidth(5);
				mScratchPaper.setIsPointNotice(false);
				break;
			case R.id.eraser:
				mPaint.setIcon(R.mipmap.icon_eraser);
				mScratchPaper.setColor(Color.WHITE);
				mScratchPaper.setStrokeWidth(40);
				mScratchPaper.setIsPointNotice(true);
				break;
			case R.id.save:
				if (mDialog!=null) {
					mDialog.dismiss();
				}
				builder = new AlertDialog.Builder(this);
				builder.setTitle(getResources().getString(R.string.save_file));
				View contentView = LayoutInflater.from(this).inflate(R.layout.include_dialog_savefile, null);
				builder.setView(contentView);
				EditText filename=(EditText) contentView.findViewById(R.id.filename);
				filename.setText(paper_name);
				builder.setNegativeButton(getResources().getString(R.string.no), new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						mDialog.dismiss();
					}
				});
				builder.setPositiveButton(getResources().getString(R.string.yes), new DialogInterface.OnClickListener() {
					@Override
					public void onClick(DialogInterface dialog, int which) {
						mDrawerLayout.closeDrawers();
						savePaper();
						mDialog.dismiss();
					}
				});
				mDialog=builder.create();
				mDialog.show();
				break;
			case android.R.id.home:
				finish();
				break;
		}
		return super.onOptionsItemSelected(item);
	}

	private void toggleFullScreen() {
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
			getWindow().getDecorView().setSystemUiVisibility(
				View.SYSTEM_UI_FLAG_LAYOUT_STABLE
				| View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION
				| View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
				| View.SYSTEM_UI_FLAG_HIDE_NAVIGATION // hide nav bar
				| View.SYSTEM_UI_FLAG_FULLSCREEN // hide status bar
				| View.SYSTEM_UI_FLAG_IMMERSIVE);
		} else {
			getWindow().setFlags(~WindowManager.LayoutParams.FLAG_FULLSCREEN,
					WindowManager.LayoutParams.FLAG_FULLSCREEN);
		}
	}

	@Override
	public void initView() {
		// TODO Auto-generated method stub
		mScratchPaper=(ScratchPaperView) findViewById(R.id.scratch_paper);
		mDrawerLayout=(DrawerLayout) findViewById(R.id.drawer_layout);
		
		mSavedPaperList=(ListView) findViewById(R.id.paper_list);
	}

	@Override
	public void initEvent() {
		// TODO Auto-generated method stub
		mSavedPaperList.setOnItemClickListener(this);
	}

	@Override
	public void initData() {
		// TODO Auto-generated method stub
		mScratchPaper.setPaperAndDesk(AppPreferenceUtils.getPaperChoose(this),
				AppPreferenceUtils.getDeskChoose(this));
		mScratchPaper.setMax_undo(AppPreferenceUtils.getMaxUndo(this));
		
		//read paper files
		mPaperList= PaperFileUtils.readPaperList();
		mAdapter=new PaperListAdapter(this,mPaperList);
		mSavedPaperList.setAdapter(mAdapter);
		
		//init paper name
		Bundle bundle=getIntent().getExtras();
		if (bundle!=null&&bundle.getString("paper_name")!=null) {
			paper_name=bundle.getString("paper_name");
			initPaperContent(paper_name);
		} else {
			paper_name= TimeUtils.getDateByFileName(Calendar.getInstance().getTimeInMillis());
		}
	}
	
	public void refreshViews() {
		mPaperList= PaperFileUtils.readPaperList();
		mAdapter=new PaperListAdapter(this,mPaperList);
		mSavedPaperList.setAdapter(mAdapter);
	}

	@Override
	protected void onPause() {
		// TODO Auto-generated method stub
		super.onPause();
		MobclickAgent.onPause(this);
	}

	@Override
	protected void onResume() {
		// TODO Auto-generated method stub
		super.onResume();
		mScratchPaper.startDraw();
		MobclickAgent.onResume(this);
	}

	@Override
	public void onClick(View v) {
		// TODO Auto-generated method stub
		super.onClick(v);
		switch (v.getId()) {
		}
	}

	@Override
	protected void onSaveInstanceState(Bundle outState) {
		// TODO Auto-generated method stub
		super.onSaveInstanceState(outState);
		Bitmap mStorePaperBackGround = Bitmap.createBitmap(mScratchPaper.getPaperWidth(),
				mScratchPaper.getPaperHeight(), Bitmap.Config.ARGB_8888);
		Canvas bitCanvas = new Canvas(mStorePaperBackGround);
		mScratchPaper.doDrawForSave(bitCanvas);
		outState.putString(KEY_STORE_BITMAP, "store_paper");
		MainApplication.store.put(outState.getString(KEY_STORE_BITMAP), mStorePaperBackGround);
		
		List<DrawStroke> mStrokeList=mScratchPaper.getStrokeList();
		outState.putString(KEY_STORE_STROKE, "store_stroke");
		MainApplication.store.put(outState.getString(KEY_STORE_STROKE), mStrokeList);
		
		mScratchPaper.stopDraw();
	}

	@Override
	protected void onRestoreInstanceState(Bundle savedInstanceState) {
		// TODO Auto-generated method stub
		super.onRestoreInstanceState(savedInstanceState);
		if (savedInstanceState.containsKey(KEY_STORE_BITMAP)) {
			mScratchPaper.setPaperBackGround((Bitmap) MainApplication.store.get(
					savedInstanceState.get(KEY_STORE_BITMAP)));
			MainApplication.store.remove(savedInstanceState.get(KEY_STORE_BITMAP));
			savedInstanceState.remove(KEY_STORE_BITMAP);
		}
		if (savedInstanceState.containsKey(KEY_STORE_STROKE)) {
			mScratchPaper.setStrokeList((List<DrawStroke>) MainApplication.store.get(
					savedInstanceState.get(KEY_STORE_STROKE)));
			MainApplication.store.remove(savedInstanceState.get(KEY_STORE_STROKE));
			savedInstanceState.remove(KEY_STORE_STROKE);
		}
		mScratchPaper.startDraw();
	}

	@Override
	public void onItemClick(AdapterView<?> parent, View view, int position,
			long id) {
		// TODO Auto-generated method stub
		AdapterView<?> mListView=parent;
		changePaperContent((String) mListView.getAdapter().getItem(position));
	}

	/**
	 * 保存草稿纸
	 */
	protected void savePaper() {
		// TODO Auto-generated method stub
		final EditText filename=(EditText) mDialog.findViewById(R.id.filename);
		if (filename.getText()==null||
				filename.getText().toString().length()==0) {
			showCustomToast(R.string.name_no_null);
			return;
		}
		
		showLoadingDialogNotCancel(R.string.saving);
		
		//启动纸张保存进程
		new Thread(new Runnable(){

			@Override
			public void run() {
				// TODO Auto-generated method stub
				if (paper_name!=null&&paper_name.length()>0) {
					PaperFileUtils.deletePaper(paper_name);
				}
				
				paper_name=filename.getText().toString();
				
				Bitmap bitmap = Bitmap.createBitmap(mScratchPaper.getPaperWidth(),
						mScratchPaper.getPaperHeight(), Bitmap.Config.ARGB_8888);
				Canvas bitCanvas = new Canvas(bitmap);
				mScratchPaper.doDrawForScreenShot(bitCanvas);
				PaperFileUtils.savePaper(bitmap, paper_name);
				ChangePaperHandler.sendEmptyMessage(0);
			}
			
		}).start();
	}
	
	protected void changePaperContent(String paper_name_tmp) {
		final String new_paper_name=paper_name_tmp;
		if (mDialog!=null) {
			mDialog.dismiss();
		}
		AlertDialog.Builder builder = new AlertDialog.Builder(this);
		builder.setTitle(getResources().getString(R.string.affirm_save_current_paper));
		View contentView = LayoutInflater.from(this).inflate(R.layout.include_change_paper, null);
		builder.setView(contentView);
		EditText editText=(EditText) contentView.findViewById(R.id.filename);
		editText.setText(paper_name);
		builder.setNegativeButton(getResources().getString(R.string.no), new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				if (new_paper_name != null) {
					paper_name = new String(new_paper_name);
				} else {
					paper_name = null;
				}
				ChangePaperHandler.sendEmptyMessage(0);
				mDialog.dismiss();
			}
		});
		builder.setPositiveButton(getResources().getString(R.string.yes), new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				final EditText filename = (EditText) mDialog.findViewById(R.id.filename);
				if (filename.getText() == null ||
						filename.getText().toString().length() == 0) {
					showCustomToast(R.string.name_no_null);
					return;
				}

				showLoadingDialogNotCancel(R.string.saving);

				//启动纸张保存进程
				new Thread(new Runnable() {

					@Override
					public void run() {
						// TODO Auto-generated method stub
						if (paper_name != null && paper_name.length() > 0) {
							PaperFileUtils.deletePaper(paper_name);
						}

						paper_name = filename.getText().toString();

						Bitmap bitmap = Bitmap.createBitmap(mScratchPaper.getPaperWidth(),
								mScratchPaper.getPaperHeight(), Bitmap.Config.ARGB_8888);
						Canvas bitCanvas = new Canvas(bitmap);
						mScratchPaper.doDrawForScreenShot(bitCanvas);
						PaperFileUtils.savePaper(bitmap, paper_name);

						if (new_paper_name != null) {
							paper_name = new String(new_paper_name);
						} else {
							paper_name = null;
						}
						ChangePaperHandler.sendEmptyMessage(0);
					}

				}).start();
				mDialog.dismiss();
			}
		});
		builder.setNeutralButton(getResources().getString(R.string.cancel), new DialogInterface.OnClickListener() {
			@Override
			public void onClick(DialogInterface dialog, int which) {
				mDialog.dismiss();
			}
		});
		mDialog=builder.create();
		mDialog.show();
	}
	
	/**
	 * 初始化纸张内容
	 * @param paper_name
	 */
	protected void initPaperContent(String paper_name) {
		if (FileUtils.isFileExist(PaperFileUtils.getPaperPath(paper_name))) {
			mScratchPaper.setPaperBackGround(PaperFileUtils.getPaper(paper_name));
			mScratchPaper.clearStrokeList();
			mScratchPaper.initPaperPosition();
		} else {
			PaperFileUtils.deletePaper(paper_name);
			finish();
		}
	}

	/**
	 * 设置LoadingDialog并显示
	 * @param resId
	 */
	protected void showLoadingDialog(int resId) {
		mLoadingDialog.setCancelable(true);
		loadingText.setText(resId);
		mLoadingDialog.show();
	}

	protected void showLoadingDialog(String message) {
		mLoadingDialog.setCancelable(true);
		loadingText.setText(message);
		mLoadingDialog.show();
	}

	protected void showLoadingDialogNotCancel(int resId) {
		mLoadingDialog.setCancelable(false);
		loadingText.setText(resId);
		mLoadingDialog.show();
	}

	protected void showLoadingDialogNotCancel(String message) {
		mLoadingDialog.setCancelable(false);
		loadingText.setText(message);
		mLoadingDialog.show();
	}

	/**
	 * 关闭LoadingDialog
	 */
	protected void dismissLoadingDialog() {
		mLoadingDialog.dismiss();
	}
}

File: android-app/app/src/main/kotlin/com/musketeer/scratchpaper/activity/EditPaperActivity.kt
package com.musketeer.scratchpaper.activity

import android.app.AlertDialog
import android.content.DialogInterface
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.v4.widget.DrawerLayout
import android.view.*
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import com.flask.colorpicker.ColorPickerView
import com.flask.colorpicker.OnColorSelectedListener
import com.flask.colorpicker.builder.ColorPickerClickListener
import com.flask.colorpicker.builder.ColorPickerDialogBuilder

import com.musketeer.scratchpaper.MainApplication
import com.musketeer.scratchpaper.R
import com.musketeer.scratchpaper.adapter.PaperListAdapter
import com.musketeer.scratchpaper.paperfile.PaperFileUtils
import com.musketeer.scratchpaper.utils.AppPreferenceUtils
import com.musketeer.scratchpaper.utils.FileUtils
import com.musketeer.scratchpaper.utils.TimeUtils
import com.musketeer.scratchpaper.view.ScratchPaperView
import com.musketeer.scratchpaper.view.ScratchPaperView.DrawStroke
import com.muskeeter.base.acitivity.BaseActivity
import com.muskeeter.base.utils.ScreenUtils
import com.nightonke.boommenu.BoomButtons.OnBMClickListener
import com.nightonke.boommenu.BoomButtons.SimpleCircleButton
import com.nightonke.boommenu.BoomMenuButton
import com.umeng.analytics.MobclickAgent
import java.util.ArrayList

import java.util.Calendar

class EditPaperActivity : BaseActivity(), OnItemClickListener, OnBMClickListener {

    companion object {
        val TAG = "EditPaperActivity"
        val ACTION_SAVE = 0
        val ACTION_SAVE_WITH_EXIT = 1
        private val KEY_STORE_BITMAP = "store_bitmap"
        private val KEY_STORE_STROKE = "store_stroke"
    }

    private var mPaint: MenuItem? = null

    private val mScratchPaper: ScratchPaperView by lazy {
        findViewById(R.id.scratch_paper) as ScratchPaperView
    }
    private var mDrawerLayout: DrawerLayout? = null
    private val mBoomMenuButton: BoomMenuButton by lazy {
        findViewById(R.id.bmb) as BoomMenuButton
    }
    private val mPaintStatus: ImageView by lazy {
        findViewById(R.id.paint_status) as ImageView
    }

    private var mDialog: AlertDialog? = null
    private var mLoadingDialog: AlertDialog? = null
    private var loadingText: TextView? = null

    //save attribute
    private var paper_name: String = ""

    private var mSavedPaperList: ListView? = null
    private var mAdapter: PaperListAdapter? = null
    private var mPaperList: MutableList<String> = ArrayList()

    private val handler = object : Handler() {
        override //当有消息发送出来的时候就执行Handler的这个方法
        fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            //只要执行到这里就关闭对话框
            refreshViews()
            dismissLoadingDialog()
            showCustomToast(resources.getString(R.string.save_success))
        }
    }

    private val ChangePaperHandler = object : Handler() {
        override //当有消息发送出来的时候就执行Handler的这个方法
        fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            if (msg.what == ACTION_SAVE_WITH_EXIT) {
                finish()
                return
            }
            //只要执行到这里就关闭对话框
            refreshViews()
            if (paper_name.isNotEmpty()) {
                initPaperContent(paper_name)
            } else {
                mScratchPaper.setPaperAndDesk(AppPreferenceUtils.getPaperChoose(this@EditPaperActivity),
                        AppPreferenceUtils.getDeskChoose(this@EditPaperActivity))
                mScratchPaper.max_undo = AppPreferenceUtils.getMaxUndo(this@EditPaperActivity)
                mScratchPaper.clearStrokeList()
                paper_name = TimeUtils.getDateByFileName(Calendar.getInstance().timeInMillis)
            }
            dismissLoadingDialog()
            mDrawerLayout!!.closeDrawer(Gravity.LEFT)
            showCustomToast(resources.getString(R.string.save_success))
        }
    }

    override fun setContentView(savedInstanceState: Bundle?) {
        val builder = AlertDialog.Builder(this)
        val view = LayoutInflater.from(this).inflate(R.layout.dialog_loading, null)
        loadingText = view.findViewById(R.id.loading_text) as TextView
        builder.setView(view)
        mLoadingDialog = builder.create()
        supportActionBar?.hide()
        setContentView(R.layout.activity_edit_paper)
    }

    override fun initView() {
        // TODO Auto-generated method stub
        mDrawerLayout = findViewById(R.id.drawer_layout) as DrawerLayout

        mSavedPaperList = findViewById(R.id.paper_list) as ListView
        mPaintStatus.setImageResource(R.drawable.ic_edit_black_24dp)
    }

    override fun initEvent() {
        // TODO Auto-generated method stub
        mSavedPaperList!!.onItemClickListener = this
        mPaintStatus.setOnClickListener(this)
    }

    fun initBmbBuilder(): SimpleCircleButton.Builder {
        return SimpleCircleButton.Builder().isRound(false).shadowCornerRadius(ScreenUtils.dpToPx(this, 20F).toInt())
                .buttonCornerRadius(ScreenUtils.dpToPx(this, 20F).toInt()).normalColor(Color.WHITE).listener(this)
    }

    override fun initData() {
        // TODO Auto-generated method stub
        mScratchPaper.setPaperAndDesk(AppPreferenceUtils.getPaperChoose(this),
                AppPreferenceUtils.getDeskChoose(this))
        mScratchPaper.max_undo = AppPreferenceUtils.getMaxUndo(this)

        //read paper files
        mPaperList = PaperFileUtils.readPaperList()
        mAdapter = PaperListAdapter(this, mPaperList)
        mSavedPaperList!!.adapter = mAdapter

        //init paper name
        val bundle = intent.extras
        if (bundle != null && bundle.getString("paper_name") != null) {
            paper_name = bundle.getString("paper_name")
            initPaperContent(paper_name)
        } else {
            paper_name = TimeUtils.getDateByFileName(Calendar.getInstance().timeInMillis)
        }

        // init boom menu
        for (i in 0..(mBoomMenuButton.buttonPlaceEnum.buttonNumber() - 1)) {
            when(i) {
                0 -> {
                    mBoomMenuButton.addBuilder(initBmbBuilder().normalImageDrawable(resources.getDrawable(R.drawable.ic_note_add_black_24dp)))
                }
                1 -> {
                    mBoomMenuButton.addBuilder(initBmbBuilder().normalImageDrawable(resources.getDrawable(R.drawable.ic_save_black_24dp)))
                }
                2 -> {
                    mBoomMenuButton.addBuilder(initBmbBuilder().normalImageDrawable(resources.getDrawable(R.drawable.ic_undo_black_24dp)))
                }
                3 -> {
                    mBoomMenuButton.addBuilder(initBmbBuilder().normalImageDrawable(resources.getDrawable(R.drawable.ic_edit_black_24dp)))
                }
                4 -> {
                    mBoomMenuButton.addBuilder(initBmbBuilder().normalImageDrawable(resources.getDrawable(R.mipmap.icon_eraser)))
                }
                5 -> {
                    mBoomMenuButton.addBuilder(initBmbBuilder().normalImageDrawable(resources.getDrawable(R.drawable.ic_clear_black_24dp)))
                }
            }
        }
    }

    override fun onBoomButtonClick(index: Int) {
        val builder: AlertDialog.Builder
        when(index) {
            0 -> {
                changePaperContent("")
            }
            1 -> {
                if (mDialog != null) {
                    mDialog!!.dismiss()
                }
                builder = AlertDialog.Builder(this)
                builder.setTitle(resources.getString(R.string.save_file))
                builder.setNegativeButton(resources.getString(R.string.no)) { dialog, which -> mDialog!!.dismiss() }
                builder.setPositiveButton(resources.getString(R.string.yes)) { dialog, which ->
                    mDrawerLayout!!.closeDrawers()
                    savePaper(ACTION_SAVE)
                    mDialog!!.dismiss()
                }
                mDialog = builder.create()
                mDialog!!.show()
            }
            2 -> {
                mScratchPaper.undoLastAction()
            }
            3 -> {
                mPaintStatus.setImageResource(R.drawable.ic_edit_black_24dp)
                if (mScratchPaper.color == Color.WHITE) {
                    mScratchPaper.color = Color.BLACK
                }
                mPaintStatus.setColorFilter(mScratchPaper.color)
                mScratchPaper.strokeWidth = 5
                mScratchPaper.isErase = false
            }
            4 -> {
                mPaintStatus.setImageResource(R.mipmap.icon_eraser)
                mPaintStatus.setColorFilter(Color.TRANSPARENT)
                mScratchPaper.color = Color.WHITE
                mScratchPaper.strokeWidth = 240
                mScratchPaper.isErase = true
            }
            5 -> {
                if (mDialog != null) {
                    mDialog!!.dismiss()
                }
                builder = AlertDialog.Builder(this)
                builder.setTitle(resources.getString(R.string.clear_all_title))
                        .setMessage(resources.getString(R.string.affirm_clear_all))
                builder.setNegativeButton(resources.getString(R.string.no)) { dialog, which -> mDialog!!.dismiss() }
                builder.setPositiveButton(resources.getString(R.string.yes)) { dialog, which ->
                    mScratchPaper.clearAll()
                    mDialog!!.dismiss()
                }
                mDialog = builder.create()
                mDialog!!.show()
            }
        }
    }

    fun refreshViews() {
        mPaperList = PaperFileUtils.readPaperList()
        mAdapter = PaperListAdapter(this, mPaperList)
        mSavedPaperList!!.adapter = mAdapter
    }

    override fun onPause() {
        // TODO Auto-generated method stub
        super.onPause()
        MobclickAgent.onPause(this)
    }

    override fun onResume() {
        // TODO Auto-generated method stub
        super.onResume()
        mScratchPaper.startDraw()
        MobclickAgent.onResume(this)
    }

    override fun onClick(v: View) {
        // TODO Auto-generated method stub
        super.onClick(v)
        when (v.id) {
            R.id.paint_status -> {
                if (mScratchPaper.isErase) {
                    return
                }
                ColorPickerDialogBuilder.with(this).setTitle(resources.getString(R.string.choose_color))
                        .wheelType(ColorPickerView.WHEEL_TYPE.FLOWER)
                        .density(12)
                        .setOnColorSelectedListener(object: OnColorSelectedListener {
                            override fun onColorSelected(selectedColor: Int) {
                                mPaintStatus.setColorFilter(selectedColor)
                                mScratchPaper.color = selectedColor
                            }
                        })
                        .setPositiveButton(resources.getText(R.string.affirm), object: ColorPickerClickListener {
                            override fun onClick(dialog: DialogInterface?, lastSelectedColor: Int, allColors: Array<out Int>?) {
                                mPaintStatus.setColorFilter(lastSelectedColor)
                                mScratchPaper.color = lastSelectedColor
                                dialog?.dismiss()
                            }
                        })
                        .setNegativeButton(resources.getText(R.string.cancel), object: DialogInterface.OnClickListener {
                            override fun onClick(dialog: DialogInterface?, which: Int) {
                                dialog?.dismiss()
                            }
                        })
                        .build().show()
            }
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        // TODO Auto-generated method stub
        super.onSaveInstanceState(outState)
        val mStorePaperBackGround = Bitmap.createBitmap(mScratchPaper.paperWidth,
                mScratchPaper.paperHeight, Bitmap.Config.ARGB_8888)
        val bitCanvas = Canvas(mStorePaperBackGround)
        mScratchPaper.doDrawForSave(bitCanvas)
        outState.putString(KEY_STORE_BITMAP, "store_paper")
        MainApplication.store.put(outState.getString(KEY_STORE_BITMAP), mStorePaperBackGround)

        val mStrokeList = mScratchPaper.strokeList
        outState.putString(KEY_STORE_STROKE, "store_stroke")
        MainApplication.store.put(outState.getString(KEY_STORE_STROKE), mStrokeList)

        mScratchPaper.stopDraw()
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        // TODO Auto-generated method stub
        super.onRestoreInstanceState(savedInstanceState)
        if (savedInstanceState.containsKey(KEY_STORE_BITMAP)) {
            mScratchPaper.paperBackGround = MainApplication.store.get(
                    savedInstanceState.get(KEY_STORE_BITMAP)) as Bitmap
            MainApplication.store.remove(savedInstanceState.get(KEY_STORE_BITMAP))
            savedInstanceState.remove(KEY_STORE_BITMAP)
        }
        if (savedInstanceState.containsKey(KEY_STORE_STROKE)) {
            mScratchPaper.strokeList = MainApplication.store.get(
                    savedInstanceState.get(KEY_STORE_STROKE)) as MutableList<DrawStroke>
            MainApplication.store.remove(savedInstanceState.get(KEY_STORE_STROKE))
            savedInstanceState.remove(KEY_STORE_STROKE)
        }
        mScratchPaper.startDraw()
    }

    override fun onItemClick(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        // TODO Auto-generated method stub
        val mListView = parent
        changePaperContent(mListView.adapter.getItem(position) as String)
    }

    /**
     * 保存草稿纸
     */
    protected fun savePaper(action : Int) {
        showLoadingDialogNotCancel(R.string.saving)

        //启动纸张保存进程
        Thread(Runnable {
            // TODO Auto-generated method stub
            if (paper_name.isNotEmpty()) {
                PaperFileUtils.deletePaper(paper_name)
            }

            val bitmap = Bitmap.createBitmap(mScratchPaper.paperWidth,
                    mScratchPaper.paperHeight, Bitmap.Config.ARGB_8888)
            val bitCanvas = Canvas(bitmap)
            mScratchPaper.doDrawForScreenShot(bitCanvas)
            PaperFileUtils.savePaper(bitmap, paper_name)
            mScratchPaper.isEdited = false
            ChangePaperHandler.sendEmptyMessage(action)
        }).start()
    }

    protected fun changePaperContent(paper_name_tmp: String) {
        val new_paper_name = paper_name_tmp
        if (mDialog != null) {
            mDialog!!.dismiss()
        }
        val builder = AlertDialog.Builder(this)
        builder.setTitle(resources.getString(R.string.affirm_save_current_paper))
        builder.setNegativeButton(resources.getString(R.string.no)) { dialog, which ->
            paper_name = new_paper_name
            ChangePaperHandler.sendEmptyMessage(0)
            mDialog!!.dismiss()
        }
        builder.setPositiveButton(resources.getString(R.string.yes), DialogInterface.OnClickListener { dialog, which ->
            showLoadingDialogNotCancel(R.string.saving)

            //启动纸张保存进程
            Thread(Runnable {
                // TODO Auto-generated method stub
                if (paper_name.isNotEmpty()) {
                    PaperFileUtils.deletePaper(paper_name)
                }

                val bitmap = Bitmap.createBitmap(mScratchPaper.paperWidth,
                        mScratchPaper.paperHeight, Bitmap.Config.ARGB_8888)
                val bitCanvas = Canvas(bitmap)
                mScratchPaper.doDrawForScreenShot(bitCanvas)
                PaperFileUtils.savePaper(bitmap, paper_name)
                paper_name = new_paper_name
                ChangePaperHandler.sendEmptyMessage(0)
            }).start()
            mDialog!!.dismiss()
        })
        builder.setNeutralButton(resources.getString(R.string.cancel)) { dialog, which -> mDialog!!.dismiss() }
        mDialog = builder.create()
        mDialog!!.show()
    }

    /**
     * 初始化纸张内容
     * @param paper_name
     */
    protected fun initPaperContent(paper_name: String) {
        if (FileUtils.isFileExist(PaperFileUtils.getPaperPath(paper_name))) {
            mScratchPaper.paperBackGround = PaperFileUtils.getPaper(paper_name)
            mScratchPaper.clearStrokeList()
            mScratchPaper.initPaperPosition()
        } else {
            PaperFileUtils.deletePaper(paper_name)
            finish()
        }
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent?): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK && mScratchPaper.isEdited) {
            mDialog?.dismiss()
            val builder = AlertDialog.Builder(this)
            builder.setTitle(resources.getString(R.string.exit_with_save_file))
            builder.setNeutralButton(resources.getString(R.string.cancel)) { dialog, which -> mDialog!!.dismiss() }
            builder.setNegativeButton(resources.getString(R.string.no)) { dialog, which ->
                mDialog?.dismiss()
                finish()
            }
            builder.setPositiveButton(resources.getString(R.string.yes)) { dialog, which ->
                mDrawerLayout!!.closeDrawers()
                savePaper(ACTION_SAVE_WITH_EXIT)
                mDialog?.dismiss()
            }
            mDialog = builder.create()
            mDialog?.show()
            return true
        }
        return super.onKeyDown(keyCode, event);
    }

    /**
     * 设置LoadingDialog并显示
     * @param resId
     */
    protected fun showLoadingDialog(resId: Int) {
        mLoadingDialog!!.setCancelable(true)
        loadingText!!.setText(resId)
        mLoadingDialog!!.show()
    }

    protected fun showLoadingDialog(message: String) {
        mLoadingDialog!!.setCancelable(true)
        loadingText!!.text = message
        mLoadingDialog!!.show()
    }

    protected fun showLoadingDialogNotCancel(resId: Int) {
        mLoadingDialog!!.setCancelable(false)
        loadingText!!.setText(resId)
        mLoadingDialog!!.show()
    }

    protected fun showLoadingDialogNotCancel(message: String) {
        mLoadingDialog!!.setCancelable(false)
        loadingText!!.text = message
        mLoadingDialog!!.show()
    }

    /**
     * 关闭LoadingDialog
     */
    protected fun dismissLoadingDialog() {
        mLoadingDialog!!.dismiss()
    }
}
